website/content/docs/rfcs/efficient-search-components.mdx

```diff
@@ -0,0 +1,341 @@
+
+The current implementation uses a simple string-based key for each node, which is sufficient for the initial use case. However, this could be extended to support more complex key structures in the future. Strings keys has the limitation of not ordering integer values.
+
+### Concurrency Control
+
+The current implementation uses tree-level locking with a single `RWMutex` per B+ tree instance. This approach was chosen for its simplicity and correctness guarantees, but it can become a performance bottleneck under high concurrent workloads.
+
+**Alternative Considered: Per-Node Locking**
+
+A more sophisticated approach would implement fine-grained locking at the node level, where each node maintains its own mutex. This would allow multiple concurrent operations to proceed simultaneously as long as they operate on different parts of the tree.
+
+```go
+type Node struct {
+    // ... existing fields
+    lock sync.RWMutex
+}
+```
+
+The tree-level locking approach was selected for the initial implementation to ensure correctness and maintainability. Per-node locking could be considered as a future optimization if benchmarking identifies concurrency as a bottleneck.
+
+### Storage
+
+The current storage system has the limitation of a one-to-one mapping between keys and nodes. This means that each key can only have one node associated with it, which can be a bottleneck when dealing with large amounts of data. To overcome this limitation, a new storage system that supports multiple nodes per key should be implemented.
+
+**Alternative Considered: Page-Based Storage**
+A page-based storage system would allow multiple nodes to be associated with a single key, enabling more efficient storage and retrieval of data. This would require a more complex implementation but would provide better performance for large datasets. The current implementation uses a simple key-value store, which is sufficient for the initial use case but may need to be replaced in the future as the system scales.
+
+## Downsides
+
+Code complexity and additional load on the system and storage when building indexes. It is suggested to make index building opt-in and also tunable so that system administrators can disable index building for mounts if desired.
+
+## Security Implications
+
+No security implications were identified.
+
+## User/Developer Experience
+
+With these constructs available, an OpenBao developer will be able to dynamically configure and build indexes on the data persisted in storage.
+No user facing changes are expected with this RFC as it only introduces the constructs to build the indexes.
+
+## Unresolved Questions
+
+- How to keep track of the current progress of an index that is being incrementally built;
+- Purging or removing all nodes associated with a key from multiple trees;
+- Index rebuilding if any configuration is changed, such as order (currently not supported); Might be easier to simply rebuild the index from scratch.
+- Removal of an index;
```

Comment:
I think the `ClearKeyView(...)` helper will be useful here (if I got the name correct from memory).